Add the End of Life (EOL) plugin for Joomla! 3.10

softforge opened the pull request:
Pull Request for Issue # .

### Summary of Changes
Updated language files
Updated quicklink plug to add new conditions, add a quicklinks group and restore the warning notice

### Testing Instructions
use a 3.9 site and set the updater to custom url. Add the github url and see that the update works and that the plugin is updated
Do a fresh install from the github install and see that the plugin is there.

### Actual result BEFORE applying this Pull Request
The Quicklinks was not in its own group but more hidden. 
There was a red image which has been replaced with the icon for information/warning
No warning across the top

### Expected result AFTER applying this Pull Request
The warning message across the top has been restored as in Joomla 2.5 EOS plugin to make it more obvious and allow a place to show different messages over time if needed
The Quicklinks is in its own group and has the icon added and the warning class added.

If the date has passed and the version is 3.10 and its the administrator control panel you should see this


You can simulate this by editing plugins->quicklink->eosnotify->eosnotify.php
 line 52 in your instillation 
if (!$this->app->isClient('administrator') || version_compare(JVERSION, '3.10', '>=') || Factory::getDate()->toSql() <= '2021-04-15')
and change the date 2021-04-15 to 2021-01-01
to make the plugin think we have passed the nominated date (my birthday) 

![eos](https://user-images.githubusercontent.com/1078933/103467087-40308000-4d43-11eb-8473-afe995204dec.PNG)


### Documentation Changes Required

@zero-24
Please take a look at this draft which I can then add to depending on the outcome of our meeting.

@bayareajenn
Was this more what you had in mind and do we know where the documentation will be held that it needs to link to?

Commit:
Add EOS Folder

Commit:
Moving EOL to Quickicons

softforge committed:
change version

softforge committed:
Add icons

softforge committed:
Add SQL files
Edit install files
Update the plugin to the latest version
Ponder Benjamins "jokes"

brianteeman commented:
We really should follow best practice and NOT have links that are just called "here".

Can I suggest something like
```
PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_MESSAGE_START=Support for Joomla 3.10 has ended. Please visit
PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_CLICK_WORD="www.joomla.org"
PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_MESSAGE_END="for more information."
```


In addition we are not really providing a good service to our users if we only tell them on the day that support has ended. I would suggest having a progressive series of messages that go from Information to Warning to Error and preferably with an actual date so a user can see something like "will end in 90 days on 1 Nov 2023". I _think_ the code for the php version eol will probably provide you with the suitable code to adopt.

brianteeman commented on plugins/quickicon/eosnotify/eosnotify.xml:
```diff
@@ -0,0 +1,20 @@
+	<authorEmail>admin@joomla.org</authorEmail>
+	<authorUrl>www.joomla.org</authorUrl>
+	<version>3.0.0</version>
+	<description>LG_QUICKICON_EOSNOTIFIER_XML_DESCRIPTION</description>
```
typo
```suggestion
	<description>PLG_QUICKICON_EOSNOTIFIER_XML_DESCRIPTION</description>
```

brianteeman commented on plugins/quickicon/eosnotify/eosnotify.xml:
```diff
@@ -0,0 +1,20 @@
+	<description>LG_QUICKICON_EOSNOTIFIER_XML_DESCRIPTION</description>
+	<files>
+		<filename plugin="eosnotify">eosnotify.php</filename>
+		<filename>index.html</filename>
```
This file is not included in the PR so can be removed from the xml. (We don't need it anyway)

brianteeman commented on plugins/quickicon/eosnotify/eosnotify.xml:
```diff
@@ -0,0 +1,20 @@
+	<license>http://www.gnu.org/licenses/gpl-2.0.html GNU/GPL</license>
+	<authorEmail>admin@joomla.org</authorEmail>
+	<authorUrl>www.joomla.org</authorUrl>
+	<version>3.0.0</version>
```
```suggestion
	<version>3.10.0</version>
```

brianteeman commented on plugins/quickicon/eosnotify/eosnotify.xml:
```diff
@@ -0,0 +1,20 @@
+	<author>Joomla! Project</author>
+	<creationDate>Jan 2021</creationDate>
+	<copyright>Copyright (C) 2005 - 2021 Open Source Matters. All rights reserved.</copyright>
+	<license>http://www.gnu.org/licenses/gpl-2.0.html GNU/GPL</license>
```
```suggestion
	<license>GNU General Public License version 2 or later; see LICENSE.txt</license>
```

softforge commented:
Thanks. Its a draft as we have a meeting in the week to decide on many aspects of the plugin including whether we have dates and whether we want all the message areas. I have updated many aspects of what was there but left the url and text as is. So please dont go too far as this will all change after the meeting in the week

brianteeman commented on plugins/quickicon/eosnotify/eosnotify.xml:
```diff
@@ -0,0 +1,20 @@
+	<name>plg_quickicon_eosnotifier</name>
+	<author>Joomla! Project</author>
+	<creationDate>Jan 2021</creationDate>
+	<copyright>Copyright (C) 2005 - 2021 Open Source Matters. All rights reserved.</copyright>
```
```suggestion
	<copyright>(C) 2021 Open Source Matters, Inc.</copyright>
```

brianteeman commented on plugins/quickicon/eosnotify/eosnotify.php:
```diff
@@ -0,0 +1,74 @@
+	 * @return array A list of icon definition associative arrays, consisting of the
+	 *				 keys link, image, text and access.
+	 *
+	 * @since       3.10
```
For consistency with references above it should be 3.10.0 (or they need to be changed to 3.10

Better yet is to do it the correct way and use __DEPLOY_VERSION__

softforge replied:
Can you please explain the "Better yet is to do it the correct way and use DEPLOY_VERSION"
I have searched through the project code and found 2 instances this being one
			if ($relativePath !== '/build/bump.php' && preg_match('#__DEPLOY_VERSION__#', $fileContents))
And I don't think that's what you are referring to. 
Do we have a doc or anything you can point to on how I would use it the correct way, keen to learn

brianteeman replied:
Firstly my apologies as github swallowed part of the code.

Instead of putting a version number that something is introduced it is best to use the string `__DEPLOY_VERSION__`

When the scripts are run to build a release it replaces that with the current version. This is particularly useful if the PR takes a while before it is merged etc so we recommend that this is always done instead of putting a specific version number in new code.

You found the correct file - documented in part here
https://docs.joomla.org/Joomla:Release_procedure_and_checklist#Building_a_Release

The code is just a few lines down from the one you quoted which basically will look for any instance of `__DEPLOY_VERSION__` and replaces it with the exact version number that is being built


https://github.com/joomla/joomla-cms/blob/81b80d03b9c80a47d4363b8c4c0fa5ec160fc1c2/build/bump.php#L331-L337

brianteeman replied:
Also if you look a few lines further down you will see that a list of all files that have updates because of this `__DEPLOY_VERSION__` magic is produced by the script.

softforge replied:
Ahh hence why I could not find more than the 2 instances, they are substituted at build, got it, and I wondered if I was barking up the wrong tree. Now I read the code and see the preg replace I see what it's doing. Thanks very helpful

brianteeman replied:
Usually I would have explained what it meant when you submitted the PR but I just assumed that you would know. My bad for assuming

brianteeman commented on plugins/quickicon/eosnotify/eosnotify.php:
```diff
@@ -0,0 +1,74 @@
+		}
+
+		return array(array(
+			'link' => 'http://www.google.com',
```
???

softforge replied:
We dont have the page yet. As I explained, its a draft to suggest how it should look so we can discuss in the CMS release meeting in the week with Tobas.  Its closer to whats needed than the old one that was in the code but will all change at the end of the week and I will then hopefully have all thats needed to complete this. I will take into account all your suggestions.

brianteeman replied:
Shame that your pr didnt include that ;)

softforge replied:
Well it is my first Brian so have to get a fair but wrong. Now I know how others feel when they submit their first ;)

brianteeman replied:
Wow - your first PR - I am truly surprised.

softforge committed:
Add icons

softforge committed:
Use __DEPLOY_VERSION__

infograf768 commented:
The Warning is not language-aware at all. A sprintf should be used.

A unique string is necessary to cope with that:

`PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_MESSAGE="Joomla 3.10 Support Has Ended. Click <a %s>Here</a> For More Information."`

and therefore delete the 3 strings:
```
PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_MESSAGE_START="Joomla 3.10 Support Has Ended Click"
PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_CLICK_WORD="Here"
PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_MESSAGE_END="for More Information."
```

Code:

`$this->app->enqueueMessage(Text::sprintf('PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_MESSAGE', 'href="https://www.joomla.org/" target="_blank"'), warning);`

which gives correctly:

<img width="458" alt="Screen Shot 2021-01-04 at 12 40 48" src="https://user-images.githubusercontent.com/869724/103531744-3d8c7280-4e8a-11eb-8d07-549c59fa0a29.png">


Also It looks like this new string is not used.
`PLG_QUICKICON_EOSNOTIFY_MESSAGE="Joomla 3.10 Support Has Ended!!!<br />Click Here for More Information."`

bayareajenn commented:
@softforge  Congratulations on your first PR! Thanks for your service!

The link will go to the same place as the last EOS plugin... docs.joomla.org/Why_Migrate. 
As for what I had in mind, I'm going to hold off on answering that part until after the CMS Release Team meets about it this week.

wilsonge commented on plugins/quickicon/eosnotify/eosnotify.php:
```diff
@@ -0,0 +1,74 @@
+		}
+
+		return array(array(
+			'link' => 'http://www.google.com',
```
```suggestion
			'link' => 'https://docs.joomla.org/Planning_for_Mini-Migration_-_Joomla_3.10.x_to_4.x',
```

HLeithner committed:
Update plugins/quickicon/eosnotify/eosnotify.php

Co-authored-by: George Wilson <georgejameswilson@googlemail.com>

bayareajenn commented:
@wilsonge  if we send them directly to the 3.10.x to 4.x planning page, then they won't see the self assessment. Or we can add a link to the self assessment to that page. I'll make a comment on the doc in Drive.

wilsonge commented on plugins/quickicon/eosnotify/eosnotify.php:
```diff
@@ -0,0 +1,74 @@
+		}
+
+		return array(array(
+			'link' => 'https://docs.joomla.org/Planning_for_Mini-Migration_-_Joomla_3.10.x_to_4.x',
```
```suggestion
			'link' => 'https://docs.joomla.org/Special:MyLanguage/Planning_for_Mini-Migration_-_Joomla_3.10.x_to_4.x',
```

bayareajenn replied:
I don't think that's a change really. But it's an easy workaround. I commented on our doc for your feedback.

wilsonge committed:
Update plugins/quickicon/eosnotify/eosnotify.php

wilsonge commented:
@bayareajenn This is true. I guess I was trying to avoid them having to click on the 3.x to 4.x funnel because in this case we should know exactly what version they are going between. I'm not sure. If you think it's better to send them to the main funnel page I'm fine with that too

sandewt commented on plugins/quickicon/eosnotify/eosnotify.php:
```diff
@@ -0,0 +1,74 @@
+use Joomla\CMS\Language\Text;
+
+/**
+ * Joomla! udpate notification plugin
```
udpate = update

bayareajenn commented:
> 
> 
> @bayareajenn This is true. I guess I was trying to avoid them having to click on the 3.x to 4.x funnel because in this case we should know exactly what version they are going between. I'm not sure. If you think it's better to send them to the main funnel page I'm fine with that too

Leaving it to link to the main funnel just removes some risk that they miss some info in those initial pages.

sandewt commented on plugins/quickicon/eosnotify/eosnotify.xml:
```diff
@@ -0,0 +1,19 @@
+<extension version="3.10" type="plugin" group="quickicon" method="upgrade">
+	<name>plg_quickicon_eosnotifier</name>
+	<author>Joomla! Project</author>
+	<creationDate>Jan 2021</creationDate>
```
`<creationDate>Jan 2021</creationDate>`
should be
`<creationDate>January 2021</creationDate>`

softforge replied:
Thank you. I am learning the ins adn outs slowly

softforge committed:
Update eosnotify.php

softforge committed:
Update the suggestions made in GitHub
Spelling corrections
Full month name
Change the message layout to use Sprint
Change copyright details

Quy commented on plugins/quickicon/eosnotify/eosnotify.php:
```diff
@@ -0,0 +1,71 @@
+		}
+
+		if ($this->app->input->get('option') == 'com_cpanel')
+		{
+			$this->app->enqueueMessage(Text::sprintf('PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_MESSAGE', 'href="https://www.joomla.org/" target="_blank"'), 'warning');
+
+		}
```
```suggestion
		{
			$this->app->enqueueMessage(Text::sprintf('PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_MESSAGE', 'href="https://www.joomla.org/" target="_blank"'), 'warning');
		}
```

sandewt commented on plugins/quickicon/eosnotify/eosnotify.php:
```diff
@@ -0,0 +1,71 @@
+		if ($this->app->input->get('option') == 'com_cpanel')
+		{
+			$this->app->enqueueMessage(Text::sprintf('PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_MESSAGE', 'href="https://www.joomla.org/" target="_blank"'), 'warning');
+
```
delete empty line 60

zero-24 replied:
```suggestion
```

sandewt commented on plugins/quickicon/eosnotify/eosnotify.php:
```diff
@@ -0,0 +1,71 @@
+	{
+		if (!$this->app->isClient('administrator') || version_compare(JVERSION, '3.10', '>=') || Factory::getDate()->toSql() <= '2021-04-15')
+		{
+			return array();
```
General comment: which code is preferred?
`return array()`
or
`return [];`

Nothing can be found in the manual: [https://developer.joomla.org/coding-standards/php-code.html](https://developer.joomla.org/coding-standards/php-code.html)

richard67 replied:
@sandewt We still have to support PHP 5.3 with Joomla 3.10, but the short array syntax `[...]` requires at least PHP 5.4, so we have to use `array()` here. In J4 we can use ` return []`.

sandewt replied:
@richard67 Thanks, for my useful information

softforge replied:
Thank you, good to know

richard67 committed:
Merge branch '3.10-dev' into 3.10/eol_plugin

richard67 commented on plugins/quickicon/eosnotify/eosnotify.php:
```diff
@@ -0,0 +1,71 @@
+
+		if ($this->app->input->get('option') == 'com_cpanel')
+		{
+			$this->app->enqueueMessage(Text::sprintf('PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_MESSAGE', 'href="https://www.joomla.org/" target="_blank"'), 'warning');
```
```suggestion
			$this->app->enqueueMessage(
				Text::sprintf(
					'PLG_QUICKICON_EOSNOTIFY_CLICK_FOR_INFORMATION_MESSAGE',
					'href="https://www.joomla.org/" target="_blank"'
				),
				'warning'
			);
```
Fix PHP code style error "Line exceeds 150 characters; contains 163 characters" reported by Drone here: https://ci.joomla.org/joomla/joomla-cms/39971/1/3

softforge replied:
Its been made shorter by the use of indexed array in my working model but your style is much more readable so will changed it even though shorter now. Thanks

richard67 commented:
@softforge You have marked my 2nd review about the code style fix as resolved, but you haven't committed that suggestion or any other fix for the code style error, and so drone is still in failed status.

brianteeman commented:
What needs to be done to complete this. 

I am assuming that this needs to be merged into 3.10 and that makes it a release blocker for 4.0 at the same time.

softforge commented:
As it needed the release dates added to make it work it was finished privately and I gave Tobias the finished package about 3-4 weeks ago. He just needs to review and add the comments the testing team made.

zero-24 commented:
Yes its still on my to do to make a PR.

brianteeman commented:
so can this one be closed?

softforge commented:
I dont see why not. Tobias, is there any reason to keep it open? If no then please close

zero-24 commented:
https://github.com/joomla/joomla-cms/pull/34453
